server/src/main/java/org/opensearch/action/search/SearchPhaseController.java

```diff
@@ -600,6 +604,42 @@ private static void validateMergeSortValueFormats(Collection<? extends SearchPha
         }
     }
 
+    /**
+     * It is necessary to widen the SortField.Type to maximum byte size for merging sorted docs.
+     * Different indices might have different types. This will avoid user to do re-index of data
+     * in case of mapping field change for newly indexed data.
+     * This will support Int to Long and Float to Double.
+     * Earlier widening of type was taken care in IndexNumericFieldData, but since we now want to
+     * support sort optimization, we removed type widening there and taking care here during merging.
+     * More details here https://github.com/opensearch-project/OpenSearch/issues/6326
+     */
+    private static void applySortFieldWidening(Sort sort, TopFieldDocs[] topFieldDocs) {
```

Comment:
@gashutos I think we could do a bit better job here by creating the `Sort` instance with widening sort fields instead of post modifying it, quick example here:

```
final Sort sort = createSort(firstTopDocs.fields);
```

```
private static Sort createSort(SortField[] fields TopFieldDocs[] topFieldDocs) {
    final SortField[] newFields = new SortField[fields.size()];

    for (int i = 0; i < fields.length; i++) {
            if (isSortWideningRequired(topFieldDocs, i) && fields[i] instanceof SortedNumericSortField) {
                final SortedNumericSortField delegate = (SortedNumericSortField) sort.getSort()[i];
                newFields[i] = new SortedWiderNumericSortField(
                    delegate.getField(),
                    delegate.getNumericType(),
                    delegate.getReverse(),
                    delegate.getSelector()
                );
            } else {
               newFields[i] = fields[i]
        }

       return new Sort(newFields);
}
```

Wdyt?

Reply:
Perfectly makes sense. I like the readibility after this change